Video 4 Zephyr - Video API, Drivers and Samples

loicpoulain opened the pull request:
DEPENDS on HAL_NXP change: https://github.com/zephyrproject-rtos/hal_nxp/pull/6

PRESENTATION: https://docs.google.com/presentation/d/1j44YHUqynN-Vw67NgHHx741yzLIbjbc76tcObi8J3Zc/edit?usp=sharing

- Generic Video API
- MCUX CSI Video driver (CMOS Sensor Interface) => Video Capture driver
- MT9M114 Video camera driver => Video control driver
- SW Video pattern generator => Video Capture driver
- A basic capture sample application, counting received frames
- A tpcserversink sample pushing video to a TCP client

Commit:
drivers: Add video API

This generic video API can be used to capture/output video frames.

Once a video buffer is enqueued to a video device endpoint, device owns
the buffer and can process to capture (camera), output (disk, display),
convert (hw encoder)... User can then call dequeue to retrieve
the processed buffer (video driver ensure cache coherency).

Once dequeued, video buffer is owned by user (e.g. for frame
processing, display buffer update, write to media, etc...).

For each video-buffer, user needs allocate the associated frame buffer
via video_buffer_alloc. Buffer format is defined by video device
endpoint configuration. Video device can be controlled (e.g. contrast,
brightness, flip...) via controls.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

Commit:
drivers: video: Add MCUX CSI video driver

Add support for CMOS Sensor Interface video driver.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

Commit:
soc: nxp: rt: Add CSI video support

Add CSI node to generic nxp rt dtsi.
Add corresponding dts binding.
Add CSI capability for rt MCUs.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

Commit:
boards: arm: nxp: mimxrt1064_evk: Enable CSI video support

This enables CSI node, and configures pinmux when driver is enabled.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

Commit:
samples: video: Add video capture sample

Simple video sample getting frames from video capture device.
Tested with mimxrt1064_evk and MT9M114 sensor.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

Commit:
CODEOWNERS: Add drivers/video path

Add myself to owner of drivers/video.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

Commit:
drivers: video: Add Aptina MT9M114 driver

MT9M114 is a CMOS digital image sensor.
Implement video interface.
Only VGA (640x480) supported for now.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

Commit:
boards: mimxrt1064_evk: Add MT9M114 digital image sensor node

Sensor can be connected to the 24pin camera connector.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

loicpoulain commented:
Hi all, any comments on this PR?

microbuilder commented:
> Hi all, any comments on this PR?

I think with the 4th of July weekend and summer holidays a lot of people are out. I'll try to have a look at this Monday myself, though.

gon1332 commented on include/drivers/video.h:
```diff
@@ -0,0 +1,317 @@
+/**
+ * @typedef video_api_get_format_t
+ * @brief get current video format
+ * See @a video_api_set_fmt() for argument.
```
`video_api_set_fmt()` to `video_api_get_fmt()`

loicpoulain replied:
thanks fixed.

gon1332 commented:
That's the first time I see this API, so I don't think I can comment more, until I actually try to port my camera module to it.
So it looks good to me for now.

pizi-nordic commented on include/drivers/video.h:
```diff
@@ -0,0 +1,317 @@
+	size_t len;
+
+	/** Flags for the buffer */
+	u8_t flags;
```
What kind of flags we have here?

loicpoulain replied:
replaced with buffer type, can be CAPTURE=empty buffer OUTPUT=filled buffer...

loicpoulain replied:
This has been removed, endpoint id defines this.

pizi-nordic commented on include/drivers/video.h:
```diff
@@ -0,0 +1,317 @@
+	void *frame;
+
+	/** Length of frame buffer in bytes */
+	size_t len;
```
This is length of buffer or length of data in the buffer?

loicpoulain replied:
This is the total len of allocated buffer.

pizi-nordic commented on include/drivers/video.h:
```diff
@@ -0,0 +1,317 @@
+ *
+ * @retval bytes per pixel
+ */
+static inline unsigned int video_pix_fmt_bpp(u32_t pixelformat)
```
Is this API valid for YUV data with color subsampling?

pizi-nordic reviewed (commented):
For me there are few blank areas in this API definition. I will try to highlight them using he following example: 

[ Camera ] => [ Software scaler/filter/color space converter ] => [ HW Encoder ] => Storage

Assumptions:
- All components in the brackets could (and should) be represented using this API.

Questions:
- Each of the device provides its own video frame allocator. Am I supposed to reallocate frame and copy its contents manually between each processing step or I can pass the frame directly?
- I assume that I enqueue uncompressed frame to the HW encoder and dequeue the compressed one. Since most of the video compression algorithms do not support compression in place, the HW encoder will allocate the output frame. Who is responsible for freeing the input frame? If user of the API, how he/she could know when the frame is no longer used. If the HW encoder driver, how it could know which driver is owner of the frame and which version of video_api_release_frame_t should be called?
- Some of the video encoders need some kind of out of band data describing properties of the compressed stream (the H264 is good example). How we are supposed to retrieve this data from the encoder and pass it to the decoder? Will we have concept of the stream in the API?

dbkinder reviewed (changes requested):
Sample and API and need documentation  (doxygen comments are a good start, but we need to get them included in the API documentation set via a .rst file, maybe in this area: https://docs.zephyrproject.org/latest/reference/peripherals/index.html

loicpoulain commented on include/drivers/video.h:
```diff
@@ -0,0 +1,317 @@
+struct video_format {
+	u32_t pixelformat;
+	u32_t width;
+	u32_t height;
```
> [ Camera ] => [ Software scaler/filter/color space converter ] => [ HW Encoder ] => Storage
> 
 
> Questions:
> 
>     * Each of the device provides its own video frame allocator. Am I supposed to reallocate frame and copy its contents manually between each processing step or I can pass the frame directly?

You only need to allocate buffer one time, and you can move them without copy/reallocation.
For now, allocate() support is optional and would be mainly implemented by  platform interface video drivers (parallel bus/CSI), because they know the low-level requirements (e.g. memory area, alignment, etc...) But we could also imagine creating a generic video allocation function (e.g. based on k_mem_pool and) which would take care of allocating buffer according to the all pipelined video device requirements (camera, hw encoder, dispaly, etc...).

>     * I assume that I enqueue uncompressed frame to the HW encoder and dequeue the compressed one. Since most of the video compression algorithms do not support compression in place, the HW encoder will allocate the output frame. Who is responsible for freeing the input frame? If user of the API, how he/she could know when the frame is no longer used. If the HW encoder driver, how it could know which driver is owner of the frame and which version of video_api_release_frame_t should be called?

You're right, Several possibilities here, but I would add a video_buffer type (e.g. CAPTURE & OUTPUT), in case of hw encoder which is a mem 2 mem device the user needs to 'queue' CAPTURE and OUTPUT buffers, the encoder converts from CAPTURE to OUTPUT buffers, on 'dequeue' CAPTURE buffers are then re-used for capture (e.g. camera, read from file...), OUTPUT buffer are the converted ones and can be e.g. sent over the network, written to SDCARD... Anyway, an enqueued buffer will always be dequeued at some point.

>     * Some of the video encoders need some kind of out of band data describing properties of the compressed stream (the H264 is good example). How we are supposed to retrieve this data from the encoder and pass it to the decoder? Will we have concept of the stream in the API?

Well, I try to start with something simple for now, but really open to improvement, adding a kind of stream context or/and some meta-data makes sense.

pizi-nordic replied:
> For now, allocate() support is optional and would be mainly implemented by platform interface video drivers (parallel bus/CSI), because they know the low-level requirements (e.g. memory area, alignment, etc...)

IMHO there should be an information about way how the video buffer is allocated inside video buffer structure. This way, any component will be able to free the buffer without knowing origin:
```
video_buffer_free(*buffer)
{
    buffer->free(buffer);
}
```

> I would add a video_buffer type like CAPTURE & OUTPUT. In case of camera user enqueue CAPTURE buffers which become OUTPUT on dequeue

I don't think that such type should exist. Buffer is just a buffer. Only HW has some limitations which must be taken under account during allocation. Ideally one allocated buffer should be used by all components in the processing pipeline. If you allow per-device buffers, you will eventually end in memory copy.

For example: The input might need buffer aligned to 4 bytes, but output device requires 16 byte alignment. If you allocate buffer from input device, you will have to copy the data. But if you allocate it from output, this will be not needed (assuming that the alignment is only problem here).

In general, I would suggest looking into OpenMAX IL and Gstreamer specifications. Both are too heavy for Zephyr, but it some problems are already solved there (like zero-copy buffer management, stream context etc.).

loicpoulain replied:
Well, I modified the API and moved video_buffer allocation from device driver to a generic video allocation handler. This simplifies the API, up to the platform/board config to overwrite default buffer alignment if requested.

loicpoulain replied:
@pizi-nordic, So I finally had some time to review different video frameworks and changed my mind on this, we can indeed see the video buffer as the transport mechanism only, without any requirement on content. The format itself is defined by the endpoint/port.

agansari replied:
I agree to this change, as a good design choice, memory should be handled in the high level system rather than the low level function.

JunYangNXP commented:
@loicpoulain @MaureenHelm @pizi-nordic @dbkinder 
With some tiny code change, I got your code running on IMXRT1050 successfully.
But could you mind to respond following comments?
1) The frame rate gets only ~6 FPS which is not as expected to reflect the performance of HW CSI and IMXRT SoC.
2) The sample is only for mt9m114 that other image sensors are hard to inherit this. I think there will be more image sensors added to support Zephyr.
3) Mt9M114 driver and CSI driver both use the same API structure video_driver_api, but actually, they are logically different devices.
4) I see mt9m114 node in the mimxrt1064_evk.dts. But actually it's not part of mimxrt1064_evk board. We can connect any other image sensors to this board.

After all, I recommend to design this video framework on top of camera, display or some other drivers. The Video framework is supposed to have functionalities to encode raw image data from camera and decode video stream (MJPEG, H.264...) to raw image data to display. This framework is supposed to be a HAL of 3th software video library or HW codec.

Please correct me if any mistake!

loicpoulain commented:
> @loicpoulain @MaureenHelm @pizi-nordic @dbkinder
 
>     1. The frame rate gets only ~6 FPS which is not as expected to reflect the performance of HW CSI and IMXRT SoC.

It's mainly because of the Sensor configuration which is applied, I copied a generic one, If you have one with correct 30, 60 or even better frame rate it would be great to integrate.

> 
>     2. The sample is only for mt9m114 that other image sensors are hard to inherit this. I think there will be more image sensors added to support Zephyr.

Yes it's explicitly for MT9M114 for now, though nothing is really specific to it in the code of the sample. Anyone adding new driver (sensor, interface, etc) can modify to make it more generic.

> 
>     3. Mt9M114 driver and CSI driver both use the same API structure video_driver_api, but actually, they are logically different devices.

They are different piece of hardware but are part of the same video 'pipeline', Making them both driven with same API abstracts this difference and low-level complexity. We can see them as both video 'sub-devices' part of the same video block, but I don't think it's necessary to add such complexity.

> 
>     4. I see mt9m114 node in the mimxrt1064_evk.dts. But actually it's not part of mimxrt1064_evk board. We can connect any other image sensors to this board.

It seems that MT9M114, as well as OV7725, have been validated with the evk board(s). But you're right it is incorrect to unconditionally add it to the DTS. What about adding the sensor node with 'status=disabled' so that we can also add the OV7725 (once enabled). This would let to the user an easy way to enable the sensor when connected.


> After all, I recommend to design this video framework on top of camera, display or some other drivers. The Video framework is supposed to have functionalities to encode raw image data from camera and decode video stream (MJPEG, H.264...) to raw image data to display. This framework is supposed to be a HAL of 3th software video library or HW codec.

Everything is possible, but again this API is here to Abstract hardware and its complexity by representing video components as IN/OUT boxes, which will have various 'controls'  (gain, exposure) depending on hardware capabilities.

loicpoulain commented on include/drivers/video.h:
```diff
@@ -0,0 +1,317 @@
+enum video_device_type {
+	video_CAPTURE_DEVICE,
+	video_OUTPUT_DEVICE,
+	video_CONTROL_DEVICE,
```
Well I removed this flag, hw encoder device will have VIDEO_CAP_M2M flag/capability.

loicpoulain replied:
This has been removed, and part of the endpoint

loicpoulain commented:
> Sample and API and need documentation (doxygen comments are a good start, but we need to get them included in the API documentation set via a .rst file, maybe in this area: https://docs.zephyrproject.org/latest/reference/peripherals/index.html

@dbkinder, done for both api/sample. I'll elaborate a bit more in API later (when final version accepted) with some schematics. (FYI presentation done during API meeting:https://docs.google.com/presentation/d/1j44YHUqynN-Vw67NgHHx741yzLIbjbc76tcObi8J3Zc/edit?usp=sharing)

loicpoulain commented:
> That's the first time I see this API, so I don't think I can comment more, until I actually try to port my camera module to it.
> So, it looks good to me for now.

@gon1332 , ok let me know if you need help, but basically, it would request to create a new uCAM video driver (either in the sample or zephyr drivers/video dir) which will read JPEG from serial to queued buffer(s), and push it for dequeue (e.g. via a FIFO) once ready.

dbkinder commented on samples/video/mt9m114/README.rst:
```diff
@@ -0,0 +1,38 @@
+Prerequisite
+************
+
+This sample requires MT9M114 to be connected.
```
connected to what?  Need to include a board reference and example of building the sample (see most any other sample for an example of using the ``.. zephyr-app-commands`` directive to document a build example).

Also looks like a serial connection is needed (to see the sample output).

loicpoulain replied:
done

dbkinder commented on samples/video/mt9m114/README.rst:
```diff
@@ -0,0 +1,38 @@
+Description
+***********
+
+This sample application simply retrieves frames from MT9M114 sensor via video
```
How about:
```
This sample application uses the Video API to retrieve video
frames from the MT9M114 sensor, writes a frame count
message to the console, and then discards the video frame data.
```

loicpoulain replied:
thanks for your help and review, change applied.

dbkinder reviewed (changes requested):
Thanks for the doc update.  Yes, after the API review adding more information about how the API is used and completing the sample documentation would be a good thing.  Coordinating with the recently proposed "camera" API would be good too since they seem to be related.

loicpoulain commented on doc/reference/peripherals/video.rst:
```diff
@@ -0,0 +1,20 @@
+.. _video:
+
+
+VIDEO
```
fixed.

dbkinder commented on samples/video/mt9m114/README.rst:
```diff
@@ -0,0 +1,59 @@
+
+The MT9M114 camera module should be plugged in the J35 camera connector on
+:ref:`mimxrt1064_evk`. A USB cable should be connected from a host to the
+micro USB debug connector (J41) in order to sget console output via the freelink
```
should ``sget`` be ``get``?

loicpoulain replied:
good catch, fixed.

dbkinder reviewed (changes requested):
One typo, otherwise doc changes LGTM

pizi-nordic commented on include/drivers/video.h:
```diff
@@ -0,0 +1,301 @@
+	void *frame;
+	u32_t bytesused;
+	u32_t type;
+	struct video_format fmt;
```
Video format will be constant in whole stream. IMHO this should be property of the stream.

loicpoulain replied:
I'm not sure about that, video buffer is a unitary element and should be self-described, and be able to exist out of a stream. E.g. If I do so post-processing on a buffer that is not part of a stream or whose stream has been released... But this also depends on how the concept of 'stream' will be implemented, If possible, I would like to go with a simple API for now, and discuss this as API improvement.

pizi-nordic replied:
That would be true if you are taking image from the camera. But video is a series of images with the same properties.

As abstraction allows for series of video frames having different format and resolution, you will have to handle that in all devices - for example by taking the values from first frame and checking in any other enqueued has the same properties.

Summarising, we have two options:
- Abstracting images (which might be later converted to video stream)
- Abstracting video (which is a stream of frames/fields).

Note, that image is unitary element, that can be processed without reference to other images. In case of video this is no longer true. While uncompressed _frames_ has this property, neither _field_ or _compressed_frame/field_  could be shown/processed  independently from the others frames/fields.

loicpoulain replied:
@pizi-nordic, So I finally had some time to review different video frameworks and changed my mind on this, we can indeed see the video buffer as the transport mechanism only, without any requirement on content. The format itself is defined by the endpoint/port.

pizi-nordic commented on include/drivers/video.h:
```diff
@@ -0,0 +1,301 @@
+	void *driver_data;
+	void *frame;
+	u32_t bytesused;
+	u32_t type;
```
I think that instead of different types of buffer (which might be CAPTURE for one component and OUTPUT for other) we should have multiple endpoints in the video devices. For in & out, there will be usually one endpoint, for M2M, two.
So you will put filled buffer to "IN" and empty to "OUT" in order to process data.

Please note that this way we can represent also things like mixers and muxers.

loicpoulain replied:
Good idea, it makes sense indeed and is much more scalable.

loicpoulain replied:
I pushed a new version and added endpoint support with basic IN and OUT eps for now.

pizi-nordic replied:
Now the API looks better.

I apologize, but in next 3 weeks I will be not able to assist you with this PR. I strongly recommend you to look into existing video APIs (for example OpenMAX and Gstreramer), as this might help you design simple and easy to use interface. For example, the endpoint idea comes directly from these frameworks.

loicpoulain replied:
Thank you for your support/ideas, I don't think there is a hurry, I'm going to look closer at OpenMAX and gstreamer designs. Note that the original version of this API is already inspired by the V4L2/Media Linux subsystem.

pizi-nordic replied:
Good. The more different API you see, the better one you create :).
I personally have to look into V4L2, as so far i worked a bit higher.

dbkinder reviewed (approved):
+1 for doc changes

loicpoulain commented:
Does anyone want to see this in 2.0 ?

agansari commented on doc/reference/peripherals/video.rst:
```diff
@@ -0,0 +1,59 @@
+   :project: Zephyr
+
+.. doxygengroup:: video_controls
+   :project: Zephyr
```
I would add the images in the ppt/presentation in this document (or the sample documentation) as an example of how-to and/or your design intention.

agansari commented on doc/reference/peripherals/video.rst:
```diff
@@ -0,0 +1,59 @@
+Video
+#####
+
+The video subsystem exposes an API to uniformly access video devices.
```
"subsystem" bothers me because it's not actually a zephyr/subsys, rather generic video driver api.

loicpoulain replied:
agree

agansari reviewed (changes requested):
NXP drivers look okay.
I have some minor change requests regarding code and documentation.

Looking at you presentation I was also expecting an API for HW encoders/decoder, not just in/out endpoints; but I guess this can be extended later on.

loicpoulain commented on boards/arm/mimxrt1064_evk/mimxrt1064_evk.dts:
```diff
@@ -75,6 +75,19 @@ arduino_i2c: &i2c1 {};
+
+		port {
+			mt9m114_ep_out: endpoint {
+				remote-endpoint = <&csi_ep_in>;
```
It's the 'official' way to link two devices though it's is currently not handled by the Zephyr devicetree parser, I expect it will be later...

loicpoulain commented:
> Looking at you presentation I was also expecting an API for HW encoders/decoder, not just in/out endpoints; but I guess this can be extended later on.

Well, a hardware encoder is just a video device with an IN and OUT endpoints for 'video buffer to be converted' and 'converted video buffer', so it can be implemented via this API. But I agree there are not yet 'controls' defined specifically for the encoder (e.g. h264 bitrate), which can be indeed added later since 'controls' is an easily extensible interface.

loicpoulain commented on drivers/video/sensor/mt9m114.c:
```diff
@@ -0,0 +1,354 @@
+		return -ENODEV;
+	}
+
+	if (val != 0x2481) {
```
fixed

loicpoulain commented on dts/bindings/video/nxp,imx-csi.yaml:
```diff
@@ -0,0 +1,30 @@
+#
+# Copyright (c) 2018 STMicroelectronics
```
fixed

loicpoulain commented on drivers/video/Kconfig.mcux_csi:
```diff
@@ -0,0 +1,11 @@
+# Kconfig - STM32 SPI driver configuration options
```
fixed

loicpoulain commented on drivers/video/Kconfig.mcux_csi:
```diff
@@ -0,0 +1,11 @@
+# Kconfig - STM32 SPI driver configuration options
+
+#
+# Copyright (c) 2015-2016 Intel Corporation
```
thanks fixed

JunYangNXP commented:
1) As we discussed, I expect to see the Video API something like power on/off, get capabilities, configuration, start/stop, en-queue/de-queue etc to cover most video instance devices.
2) Do we need to have generic driver data structure and dts bindings for video drivers?
3) Video sample is supposed to be generic rather than only mt9m114.

@MaureenHelm

loicpoulain commented:
Hey Jian,

>     1. As we discussed, I expect to see the Video API something like

> power on/off

It looks like not specifically related to video devices, but to any device. This kind of feature should be addressed at the generic device level if we want to support it. Not sure if there any plan here.

> get capabilities

I'm not against this, which capabilities do you want to report? I was assuming that, since we are in RTOS context, the developer knows a bit about the hardware and we don't want to unnecessarily increase memory footprint.

> configuration

Is done via set/get format, and via extensible 'controls' interface for other parameters.

> start/stop

Is done via start/stop stream.

> en-queue/de-queue

Is done via enqueue/dequeue.

>     2. Do we need to have generic driver data structure and dts bindings for video drivers?

The DTS bindings depend on the hardware and differ depending on the device (parallel port, spi, usb....). Do you think about something specifically?

>     3. Video sample is supposed to be generic rather than only mt9m114.

The source itself is quite generic, but not the prj.conf, I'm still thinking about the best way to do that, maybe with a config overlay.

JunYangNXP commented:
> Hey Jun,
> 
> > ```
> > 1. As we discussed, I expect to see the Video API something like
> > ```
> 
> > power on/off
> 
> It looks like not specifically related to video devices, but to any device. This kind of feature should be addressed at the generic device level if we want to support it. Not sure if there any plan here.

[Jun] We could need this for low power consumption.
> 
> > get capabilities
> 
> I'm not against this, which capabilities do you want to report? I was assuming that, since we are in RTOS context, the developer knows a bit about the hardware and we don't want to unnecessarily increase memory footprint.
> 
[Jun] User could need to know capabilities of video device something like pixel format, resolution..We can't assume user knows the details of any device and user program could be compatible with any device.
BTW, the API added require only very little memory.
> > configuration
> 
> Is done via set/get format, and via extensible 'controls' interface for other parameters.
> 
> > start/stop
> 
> Is done via start/stop stream.
> 
> > en-queue/de-queue
> 
> Is done via enqueue/dequeue.
> 
> > ```
> > 2. Do we need to have generic driver data structure and dts bindings for video drivers?
> > ```
> 
> The DTS bindings depend on the hardware and differ depending on the device (parallel port, spi, usb....). Do you think about something specifically?
> 
> > ```
> > 3. Video sample is supposed to be generic rather than only mt9m114.
> > ```
> 
> The source itself is quite generic, but not the prj.conf, I'm still thinking about the best way to do that, maybe with a config overlay.
[Jun] not only mt9m114, the board/chip name could not appear in sample code.

PS, do we need video framework to manage the frame buffer? It supposed to be handled by specific driver or user program.

loicpoulain commented:
> > Hey Jun,

> [Jun] We could need this for low power consumption.

Maybe add a specific 'ctrls' for that?


> > I'm not against this, which capabilities do you want to report? I was assuming that, since we are in RTOS context, the developer knows a bit about the hardware and we don't want to unnecessarily increase memory footprint.
> 
> [Jun] User could need to know capabilities of video device something like pixel format, resolution..We can't assume user knows the details of any device and user program could be compatible with any device.
> BTW, the API added require only very little memory.

Well we can certainly add this function.

> PS, do we need video framework to manage the frame buffer? It supposed to be handled by specific driver or user program.

What to you mean exactly? you refer to buffer allocation?

carlescufi commented:
Discussed during the API meeting. Could the rtio framework be shared for V4Z as well? See https://github.com/zephyrproject-rtos/zephyr/pull/17434

loicpoulain commented:
> Discussed during the API meeting. Could the rtio framework be shared for V4Z as well? See #17434

Well, at some point I imagine we can rebase video 4 zephyr on using RTIO (thugh I need to have a deeper look), when merged.

loicpoulain commented:
@pizi-nordic could you please review again

pizi-nordic commented:
@loicpoulain: I will try to do the review next week. Now I am a bit busy coding for Zephyr 2.0 :)

Commit:
samples: video: Add TCP server sink sample

This samples captures frames from video capture device (in any format),
and sends them to its TCP client.

Tested with:
 - mimxrt1064 + MT9M114 video sensor.
 - Gstreamer 1.8.3 running on host

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

loicpoulain commented:
> @loicpoulain: I will try to do the review next week. Now I am a bit busy coding for Zephyr 2.0 :)

No problem, in the meantime

loicpoulain commented:
> @loicpoulain: I will try to do the review next week. Now I am a bit busy coding for Zephyr 2.0 :)

No problem, in the meantime I've added a video sample to stream video over TCP. I plan to add also a test pattern software device to generate video pattern in software and test the subsystem (maybe in a future PR).

Also, NXP_HAL is a separate repo now, so the following change is requested:
https://github.com/zephyrproject-rtos/hal_nxp/pull/6

Commit:
drivers: video: Add software video pattern generator

This is a virtual device generating video pattern for testing
purpose. It supports colobar pattern for now.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

Commit:
drivers: video: Remove dedicated sensor dir

Keep flat video driver directory for now.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

loicpoulain commented:
@pizi-nordic I added a Video Software Generator virtual device/driver for testing purpose. It generates colorbar rgb565 image. This should allow testing video 'subsystem' without hw camera.

carlescufi commented:
> @loicpoulain: I will try to do the review next week. Now I am a bit busy coding for Zephyr 2.0 :)

@pizi-nordic: Any chance you can follow-up on this?

carlescufi commented:
@bfrog could you give this PR a review?

pizi-nordic commented:
@carlescufi, @loicpoulain: I will do the review this week. Most likely on Friday.

pizi-nordic commented on include/drivers/video.h:
```diff
@@ -0,0 +1,518 @@
+				   struct video_buffer **buf, u32_t timeout);
+
+/**
+ * @typedef video_api_stream_start_t
```
We have to document what will happen / what to do with the video buffers on the stream start/stop.
Questions:
- Do I need pre-fill input device with buffers before starting the stream?
- What will happen to buffers enqueued to the device after stopping the stream? If they will be returned without data, how I can distinguish processed/not-processed buffer?

loicpoulain replied:
What do you think about the following:

- video_stream_start is called to enter ‘streaming’ state; the driver may receive buffers with video_enqueue() before start_streaming is called; if driver/device needs a minimum number of buffers before being able to start streaming, then driver set the min_vbuf_count in endpoint capabilities. => TODO add min_vbuf_count to video_caps
 
- on video_stream_stop driver should stop any transactions or wait until they finish and give back all buffers, the BUF_DONE event(s) is(are) generated, and driver calls registered callback (if any). The user has to dequeue the remaining buffers with video_dequeue().

pizi-nordic replied:
Sounds good. However you can have conflicting statements in stop scenario: You have to pre-queue buffers to the halted device, but after stopping (so in the same halted state state), device returns back all buffers. So stopped device have to hold buffers unprocessed buffers as well as return them back ;-)

loicpoulain replied:
Well, I can certainly reword that, it's not the fact that the device is in stop state that make buffers available for dequeue, but the fact that calling stop() cause any currently queued buffer (meaning: owned by the driver/device) to be aborted and then dequeue able by the user. Any buffer can then be queued again and will be dequeuable only on completion (capture done...) or after a new call of stop (aborted). Does it make sense?

pizi-nordic replied:
Then in some cases you will have call stop multiple time. What about additional state (like PREPARE)? In prepare you can enqueue the buffers, but they will be not processed, START starts the processing, STOP returns all buffers.

loicpoulain replied:
I'm not sure to understand which 'normal' scenario can lead to calling stop multiple times? I think I understand your point, you link the 'returned buffer behavior' with a state (STOP) whereas I link this behavior to an event (call to stop()).  I would prefer to keep API simple and not adding extra prepare state, I think it would unnecessary complexify video device management. Instead, maybe I can add a dedicated flush() function to abort any queued buffer, but in current API the flush() is implied by stop().

loicpoulain replied:
@pizi-nordic , is there anything I can do to move forward on this point?

loicpoulain replied:
discussed with Piotr, added a flush function allowing to flush the incoming buffers either by waiting on video function completion or by canceling them all.

pizi-nordic commented on include/drivers/video.h:
```diff
@@ -0,0 +1,518 @@
+
+/**
+ * @typedef video_api_set_ctrl_t
+ * @brief Stop the capture or output process.
```
Copy/Paste error.

pizi-nordic commented on include/drivers/video.h:
```diff
@@ -0,0 +1,518 @@
+ * @param buffer is a pointer to the start of the buffer.
+ * @param bytesused is the number of bytes occupied by the valid data in
+ *        the buffer.
+ * @param fmt is the video frame format.
```
There is no fmt in the video_buffer struct.

pizi-nordic commented on include/drivers/video.h:
```diff
@@ -0,0 +1,518 @@
+
+/**
+ * @typedef video_api_get_ctrl_t
+ * @brief Stop the capture or output process.
```
Copy/Paste error.

pizi-nordic commented on include/drivers/video.h:
```diff
@@ -0,0 +1,518 @@
+
+/**
+ * @typedef video_api_register_event_cb_t
+ * @brief Register callback for video endpoint events.
```
Is the typedef below registering a callback? ;-)

pizi-nordic commented on include/drivers/video.h:
```diff
@@ -0,0 +1,518 @@
+ *
+ * @retval 0 Is successful, -ERRNO code otherwise.
+ */
+static inline int video_register_event_cb(struct device *dev,
```
How I can de-register callback?

loicpoulain replied:
Either I add the symmetric or consider passing NULL cb to register_function

pizi-nordic replied:
IMHO NULL is better

pizi-nordic commented on include/drivers/video.h:
```diff
@@ -0,0 +1,518 @@
+{
+	const struct video_driver_api *api =
+		(const struct video_driver_api *)dev->driver_api;
+
```
IMHO we should have ASSERTs checking if the backing API is implemented (for mandatory methods) or we should return -ENOSUP (if method is optional and not implemented)

loicpoulain replied:
OK, I'm going to add that and clarify which one is mandatory/optional

pizi-nordic reviewed (dismissed):
There are still holes in the API definition

loicpoulain commented on include/drivers/video.h:
```diff
@@ -0,0 +1,518 @@
+ * @param height_min is the minimum supported frame width.
+ * @param height_max is the maximum supported frame width.
+ */
+struct video_caps {
```
Well, we can probably add a 'step' parameter for this.

pizi-nordic replied:
We should also allow to link the video_caps structures. This way we will be able to represent the device which can only stream in N specific resolutions/formats. Like some web caps which streams at VGA resolution in MJPEG or QVGA in YUV.

loicpoulain replied:
done, video_caps now contains a list of format_cap

loicpoulain commented on include/drivers/video.h:
```diff
@@ -0,0 +1,518 @@
+ *        the buffer.
+ * @param fmt is the video frame format.
+ * @param timestamp time as returned by k_cycle_get_32() at which the last
+ *        data byte was actually received (input) or consumed (output).
```
yes, that makes sense.

bfrog commented on include/drivers/video.h:
```diff
@@ -0,0 +1,556 @@
+	const struct video_driver_api *api =
+		(const struct video_driver_api *)dev->driver_api;
+
+	if (api->enqueue == NULL) {
```
When would a driver not implement this?

bfrog commented on include/drivers/video.h:
```diff
@@ -0,0 +1,556 @@
+	const struct video_driver_api *api =
+		(const struct video_driver_api *)dev->driver_api;
+
+	if (api->dequeue == NULL) {
```
Same comment as above, it seems like this and enqueue should be required instead of optional

loicpoulain replied:
Well, there is some configuration implying non-bufferable video devices, such devices are 'control' only (set/get_format, set/get-ctrls...). These devices are generally slaves and controlled by a master video device (and not directly by the user). This is the case for example on nxp mimxrt1064-evk board, there are two video devices, the SoC video interface (NXP CSI), and the sensor (e.g. MT9M114) controlled via I2C, but outputting useful data to the parallel interface (CSI). In that case CSI device is the main interface the user deal with and CSI driver control internally the 'slave' which can be MT9M114 sensor or any other one.

 USER
      |
 [NXP-CSI] <-> [MT9M114-I2C-SENSOR]

bfrog commented on include/drivers/video.h:
```diff
@@ -0,0 +1,556 @@
+ *
+ * @retval pointer to allocated video buffer
+ */
+struct video_buffer *video_buffer_alloc(size_t size);
```
If allocation fails does this return a NULL? Does this block the calling context until the allocation can happen?

loicpoulain replied:
It does not block and return NULL if memory is not available.

bfrog commented on include/drivers/video.h:
```diff
@@ -0,0 +1,556 @@
+		return -ENOTSUP;
+	}
+
+	return api->dequeue(dev, ep, buf, timeout);
```
By hiding the IPC mechanism here your disallowing the use of k_poll. In the case where you have multiple cameras you'd like to recieve data from, with this API, you'd be polling each camera with this API using K_NO_WAIT, spinning. With k_poll you would get which video/camera device has a frame available as they come in. I would prefer to see something that is k_poll-able here.

bfrog commented on include/drivers/video.h:
```diff
@@ -0,0 +1,556 @@
+ * @brief Register/Unregister event callback for a video endpoint.
+ *
+ * Register a endpoint event handler which is called on frame completion,
+ * error, etc (cf video_event enum). There is no guarentee on the
```
So this answers my above comments about the context of the callback. If there is no guarantee of context the caller must assume an IRQ context which is the least useful place to get a callback as it effectively means your going to use some form of IPC or task queue to do the actual work. In which case, why not use a k_poll-able signal here instead

loicpoulain replied:
Yes I tend to agree here, what do you think about replacing register_event_callback with register_event_signal, this would allow the user to poll on the signal, and use standard dequeue function on e.g. BUFFER_DONE event.

bfrog replied:
I think that would be fine, the exception might be that you'd like to actually have code run in an ISR context. For example if it really can't be scheduled for later then using a signal would not make sense, but this would need to be clearer about that.

loicpoulain replied:
Yes, the only code we would like to run in IRQ context is for the video device management, and it's already done in the driver anyway, and there is no guarantee for the callback context (it depends on the driver). Most of the time, the event will lead to dequeue a buffer and process it, or requeue it to another video device. The only problem I see with k_poll is that we can miss events in case signals come faster than managed in user thread, which then needs to be taken into account.

loicpoulain replied:
I've added video_set_signal() allowing to register an asynchonous signal which is then signaled on video endpoint buffer event.

teburd reviewed (commented):
I generally think this is ok though I would prefer to use k_poll friendly things instead of video_dequeue and callbacks as it would allow easier usage with multiple devices.

Secondly this has the same general issue that my own API has I believe, based on my understanding of user mode, it will not work with user mode threads which I think would be a highly desired feature for video processing, decoding, and encoding work.

loicpoulain commented on include/drivers/video.h:
```diff
@@ -0,0 +1,556 @@
+ * @param user_data Pointer to user data.
+ *
+ */
+typedef void (*video_event_callback_t)(struct device *dev,
```
added signal support

carlescufi commented:
@pizi-nordic is there a chance you could take one last look at this PR?

MaureenHelm reviewed (changes requested):
Have you tried sending video to the LCD display?

loicpoulain commented on dts/bindings/video/aptina,mt9m114.yaml:
```diff
@@ -0,0 +1,14 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+title: BMI160 Inertial measurement unit
```
thanks done

loicpoulain commented on dts/arm/nxp/nxp_rt.dtsi:
```diff
@@ -546,6 +546,7 @@
 			interrupts = <111 0>;
 			clocks = <&ccm IMX_CCM_USDHC2_CLK 0 0>;
 			label = "USDHC_2";
+		};
```
yes, fixed now

loicpoulain commented:
> Have you tried sending video to the LCD display?

No, I do not have LCD for the iMX RT, but AFAIR,  @JunYangNXP knows quite well how to do this, not sure if he has time, but it would be a good end-to-end sample to add.

MaureenHelm commented:
> > Have you tried sending video to the LCD display?
> 
> No, I do not have LCD for the iMX RT, but AFAIR, @JunYangNXP knows quite well how to do this, not sure if he has time, but it would be a good end-to-end sample to add.

I'm wondering about the interaction between this video api and the existing display api in `include/drivers/display.h`. Will they work together? Would it make sense to build a video output driver on top of the display api?

There is a display driver for iMX RT in `drivers/display/display_mcux_elcdif.c`

loicpoulain commented:
> > > Have you tried sending video to the LCD display?
> > 
> > 
> > No, I do not have LCD for the iMX RT, but AFAIR, @JunYangNXP knows quite well how to do this, not sure if he has time, but it would be a good end-to-end sample to add.
> 
> I'm wondering about the interaction between this video api and the existing display api in `include/drivers/display.h`. Will they work together? Would it make sense to build a video output driver on top of the display api?

It's certainly do-able yes, the APIs are quite similar and a glue driver is possible. I'll also investigate if it would be worth to convert display API to video one. But it's a long term task.

> 
> There is a display driver for iMX RT in `drivers/display/display_mcux_elcdif.c`

Ok, I'm going to try getting one.

carlescufi commented:
> > > Have you tried sending video to the LCD display?
> > 
> > 
> > No, I do not have LCD for the iMX RT, but AFAIR, @JunYangNXP knows quite well how to do this, not sure if he has time, but it would be a good end-to-end sample to add.
> 
> I'm wondering about the interaction between this video api and the existing display api in `include/drivers/display.h`. Will they work together? Would it make sense to build a video output driver on top of the display api?
> 
> There is a display driver for iMX RT in `drivers/display/display_mcux_elcdif.c`

@vanwinkeljan could you please take a look at the interaction between this new API and the existing display API?

vanwinkeljan commented:
> @vanwinkeljan could you please take a look at the interaction between this new API and the existing display API?

@carlescufi / @MaureenHelm It looks pretty straightforward to build a glue logic video endpoint on top of the existing `display` API. But I think there is still a use case to support both APIs; to me it looks like the `video` API is intended to operate on full screen size and update the complete display at once whereas the `display` API allows to do partial updates, this feature could be of importance to memory constrained devices. Of Course the video API could be extended to support this but I would not block this PR for that.

vanwinkeljan commented on include/drivers/video.h:
```diff
@@ -0,0 +1,566 @@
+ *        output endpoints.
+ */
+struct video_buffer {
+	void *driver_data;
```
Would it no be better to change this to a struct with as first member a type identifier and as second member the data pointer?

loicpoulain replied:
Well, not sure why adding a type here, this is a data-agnostic pointer (private to the driver), each video driver can use its own way. E.g. in existing video drivers a fifo is used and this private pointer is used for fifo managment. It's not restricted to specific types.

vanwinkeljan replied:
Ok my misunderstanding, I thought it was used to communicate driver specific data between endpoints.

loicpoulain commented:
.



> > @vanwinkeljan could you please take a look at the interaction between this new API and the existing display API?
> 
> @carlescufi / @MaureenHelm It looks pretty straightforward to build a glue logic video endpoint on top of the existing `display` API. But I think there is still a use case to support both APIs; to me it looks like the `video` API is intended to operate on full screen size and update the complete display at once whereas the `display` API allows to do partial updates, this feature could be of importance to memory constrained devices. Of Course the video API could be extended to support this but I would not block this PR for that.

Thanks for your review, indeed it's easy to feed the display API with video API buffer content. Let's see later if there is a need and a way to extend video API to include display/graphics capabilities. For now we keep both and I'll submit a sample using Video API for capture + colorspace conversion and Display API for rendering.

carlescufi commented:
> > @vanwinkeljan could you please take a look at the interaction between this new API and the existing display API?
> 
> @carlescufi / @MaureenHelm It looks pretty straightforward to build a glue logic video endpoint on top of the existing `display` API. But I think there is still a use case to support both APIs; to me it looks like the `video` API is intended to operate on full screen size and update the complete display at once whereas the `display` API allows to do partial updates, this feature could be of importance to memory constrained devices. Of Course the video API could be extended to support this but I would not block this PR for that.

Thank you @vanwinkeljan for the review.
@MaureenHelm would you mind approving the PR or removing your NACK?
